src/dmd/dmodule.d

```diff
@@ -653,188 +653,199 @@ extern (C++) final class Module : Package
+                if (buf.length >= 4 && buf[1] == 0 && buf[2] == 0 && buf[3] == 0)
                 {
-                    if (buf[1] == 0 && buf[2] == 0 && buf[3] == 0)
-                    {
-                        // UTF-32LE
-                        le = 1;
-                        goto Lutf32;
-                    }
-                    else if (buf[0] == 0 && buf[1] == 0 && buf[2] == 0)
-                    {
-                        // UTF-32BE
-                        le = 0;
-                        goto Lutf32;
-                    }
+                    endian = Endian.little;
+                    sourceEncoding = SourceEncoding.utf32;
                 }
-                if (buflen >= 2)
+                else if (buf.length >= 4 && buf[0] == 0 && buf[1] == 0 && buf[2] == 0)
                 {
-                    if (buf[1] == 0)
-                    {
-                        // UTF-16LE
-                        le = 1;
-                        goto Lutf16;
-                    }
-                    else if (buf[0] == 0)
-                    {
-                        // UTF-16BE
-                        le = 0;
-                        goto Lutf16;
-                    }
+                    endian = Endian.big;
+                    sourceEncoding = SourceEncoding.utf32;
+                }
+                else if (buf.length >= 2 && buf[1] == 0) //try to check for UTF-16
+                {
+                    endian = Endian.little;
+                    sourceEncoding = SourceEncoding.utf16;
                 }
-                // It's UTF-8
-                if (buf[0] >= 0x80)
+                else if (buf[0] == 0)
                 {
-                    error("source file must start with BOM or ASCII character, not \\x%02X", buf[0]);
+                    endian = Endian.big;
+                    sourceEncoding = SourceEncoding.utf16;
+                }
+                else {
+                    // It's UTF-8
+                    needsReencoding = false;
+                    if (buf[0] >= 0x80)
+                    {
+                        error("source file must start with BOM or ASCII character, not \\x%02X", buf[0]);
                     fatal();
+                    }
                 }
             }
+            //throw away BOM
+            if (hasBOM){
+                if (!needsReencoding) buf = buf[3..$];// utf-8 already
+                else if (sourceEncoding == SourceEncoding.utf32) buf = buf[4..$];
+                else buf = buf[2..$]; //utf 16
+            }
+        }
+        //printf("%s, %d, %d, %d\n", srcfile.name.toChars(), needsReencoding, endian == Endian.little, sourceEncoding == SourceEncoding.utf16);
+        if (needsReencoding)
+        {
+            if (sourceEncoding == SourceEncoding.utf16)
+            {
+                buf = endian == Endian.little ?
+                    toUtf8!(SourceEncoding.utf16, Endian.little)(buf) :
+                    toUtf8!(SourceEncoding.utf16, Endian.big)(buf);
```

Comment:
Please don't invent a different styling.

Reply:
I tried to find an example to replicate... please let me know if I'm still off.

Reply:
The usual is:
```
a  = condition
     ? one
     : other;
```
Lining them up on the left makes it easy to see it's a ?: expression.